Issue 136

CodeCombiner opened the pull request.

CodeCombiner committed:
markdown links to html links completed, tests passed

CodeCombiner committed:
Merge branch 'master' of https://github.com/Hamstoo/data-model into issue-136

Pulling from master

CodeCombiner committed:
embedded marks wrapped

CodeCombiner committed:
fix formatting

CodeCombiner committed:
fix formatting

fcrimins commented on src/test/scala/com/hamstoo/models/MarkTests.scala:
```diff
@@ -21,8 +21,42 @@ class MarkTests extends FlatSpecWithMatchers with DataInfo {
+      "<p>hello markdown link conversion text " +
+      "[I'm markdown link](https://www.google.com)</p>"
+
+   val s1 =  MarkData.embedLinksToHtmlLinks(nonFilteredOfEmbeddedLinksTags)
```
Based on where `embedLinksToHtmlLinks` is called in `commentEncoded` it looks to me like it is supposed to operate on markdown formatted (not yet rendered to HTML) text, but this this test is passing in HTML, which is confusing.  Rather than calling `embedLinksToHtmlLinks` before rendering, which requires the identification of both markdown and HTML-tagged links, why not call it afterward (but before XSS sanitization) when only HTML-tagged links would need to be identified.  This would perhaps simplify `ignoreTagsAndFindLinksInText` which, at it looks now, is unintelligible and unmaintainable, especially given that it's not commented.

CodeCombiner replied:
Fixed according to comment:
1) Comments fixed and now describe correctly  `embeddedLinksToHtmlLinks` method;
2) `embeddedLinksToHtmlLinks` called right after rendering to improve maintenance
3) Removed particular logics from `embeddedLinksToHtmlLinks` to ignore markdown tags to avoid unnecessary filtering which is done by CommonMark parser.
4) Test fixed because now `embeddedLinksToHtmlLinks` does not notice markdown tagged links

CodeCombiner committed:
Source fixed according to comment https://github.com/Hamstoo/data-model/pull/9#discussion_r147469908

CodeCombiner committed:
added SNAPSHOT for branch

fcrimins commented on src/main/scala/com/hamstoo/models/mark.scala:
```diff
@@ -95,6 +103,16 @@ object MarkData {
+  // Find all embed urls and convert them to html <a> links (anchors)
+  // regex designed to ignore html link tag and markdown link tag
+  def embeddedLinksToHtmlLinks(text: String): String = {
+    val ignoreTagsAndFindLinksInText =
```
This regex is unmaintainable.  At the very least it requires a comment (or several comments, one for each constituent part) explaining what it's supposed to be doing.

fcrimins commented on src/main/scala/com/hamstoo/models/mark.scala:
```diff
@@ -95,6 +103,16 @@ object MarkData {
   val commentMergeSeparator: String = "\n\n---\n\n"
+
+  // Find all embed urls and convert them to html <a> links (anchors)
+  // regex designed to ignore html link tag and markdown link tag
```
This does not follow our style for function header comments.

fcrimins commented on src/test/scala/com/hamstoo/models/MarkTests.scala:
```diff
@@ -21,8 +21,39 @@ class MarkTests extends FlatSpecWithMatchers with DataInfo {
+    val nonFilteredOfEmbeddedLinksTags = "<p>hello embedded link in text " +
+      // this html link should be found and wrapped to anchor tag <a>
+      "https://www.inTextEmbeddedUntaggedLink.com </p>" +
+      // this html link should be skipped and stay as is
```
1. Split this into 2 tests, one that tests that https://www.inTextEmbeddedUntaggedLink.com gets properly tagged and another that ensures <a href=\"https://www.google.com\"> does not.
2. https://www.inTextEmbeddedUntaggedLink.com should get tagged like this: `<a href=\"https://www.inTextEmbeddedUntaggedLink.com\" rel=\"nofollow noopener noreferrer\" target=\"_blank\">`.  I.e. just like google.com is getting tagged in the previous test.

CodeCombiner replied:
Like this

```https://www.inTextEmbeddedUntaggedLink.com should get tagged like this: <a href=\"https://www.inTextEmbeddedUntaggedLink.com\" rel=\"nofollow noopener noreferrer\" target=\"_blank\">. I.e. just like google.com is getting tagged in the previous test.```

This link get tagged later in the source by this function `Jsoup.clean(html2, htmlTagsWhitelist)`

fcrimins commented on src/test/scala/com/hamstoo/models/MarkTests.scala:
```diff
@@ -21,8 +21,39 @@ class MarkTests extends FlatSpecWithMatchers with DataInfo {
+      "I'm an inline-style link</a></p>"
+
+
+   val s1 =  MarkData.embeddedLinksToHtmlLinks(nonFilteredOfEmbeddedLinksTags)
```
Improper tabbing and double empty lines of whitespace are a distraction.

CodeCombiner committed:
1) Fixed Improper tabbing and double empty lines
2) Fixed comment according to our style (Scala doc used instead of simple comment)
3) Test has been split in two
4) Regex patterns commented and improved, added 2 tests by ip and by domain

CodeCombiner commented:
The commit includes

https://github.com/Hamstoo/data-model/pull/9/commits/39534b0da3e001a384b7ea372c9965a24fd96a90

1) Fixed Improper tabbing and double empty lines
2) Fixed comment according to our style (Scala doc used instead of simple comment)
3) Test has been split in two
4) Regex patterns commented and improved, added 2 tests by ip and by domain

CodeCombiner committed:
fixed regex, added nodes walk through to improve maintenance

CodeCombiner committed:
fixed regex, added nodes walk through to improve maintenance

CodeCombiner committed:
fixed typo

CodeCombiner committed:
fixed empty spaces at the end of the code block

fcrimins reviewed (approved):
Still some minor styling inconsistencies, but looks pretty good o/w.  It might be sufficient to just check for 'href' and protocol://<non-whitespace> and remove domain name regex, domain zone regex, and allowed symbols regex.  The downside of a false positive that we tag as a link doesn't seem that large.  Please go ahead and merge into master branch.

CodeCombiner committed:
allowed anything which is allowed in url, removed domain and zone check

CodeCombiner committed:
fix version conflict, remove unnecessary comment
